examples/random.h

```diff
@@ -0,0 +1,54 @@
+/**********************************************************************
+ * Copyright (c) 2020 Elichai Turkel	                              *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+/*
+ * In this file only the best practice randomness sources are used
+ * Platform randomness sources:
+ * Linux   -> `getrandom(2)`(`sys/random.h`), if not available `/dev/urandom` should be used. http://man7.org/linux/man-pages/man2/getrandom.2.html, https://linux.die.net/man/4/urandom
+ * macOS   -> `getentropy(2)`(`sys/random.h`), if not available `/dev/urandom` should be used. https://www.unix.com/man-page/mojave/2/getentropy, https://opensource.apple.com/source/xnu/xnu-517.12.7/bsd/man/man4/random.4.auto.html
+ * FreeBSD -> `getrandom(2)`(`sys/random.h`), if not available `kern.arandom` should be used. https://www.freebsd.org/cgi/man.cgi?query=getrandom, https://www.freebsd.org/cgi/man.cgi?query=random&sektion=4
+ * OpenBSD -> `getentropy(2)`(`unistd.h`), if not available `/dev/urandom` should be used. https://man.openbsd.org/getentropy, https://man.openbsd.org/urandom
+ * Windows -> `BCryptGenRandom`(`bcrypt.h`). https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom
+ */
+
+#if defined(_WIN32)
+#include <bcrypt.h>
+#elif defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)
+#include <sys/random.h>
+#elif defined(__OpenBSD__)
+#include <unistd.h>
+#else
+#error "Couldn't identify the OS"
+#endif
+
+
+/* Returns 1 on success, and 0 on failure. */
+static int fill_random(unsigned char* data, unsigned long size) {
+#if defined(_WIN32)
+    NTSTATUS res = BCryptGenRandom(NULL, data, size, BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+    if (res == STATUS_SUCCESS) {
+        return 1;
+    } else {
+        return 0;
+    }
+#elif defined(__linux__) || defined(__FreeBSD__)
+    /* If `getrandom(2)` is not available you should fallback to /dev/urandom */
+    ssize_t res = getrandom(data, size, 0);
+    if (res == (ssize_t)size) {
```

Comment:
`unsigned long` may not fit in an `ssize_t`. On 32 bit systems it can therefore happen that getrandom returns -1 but `fill_random` returns `1`. Perhaps change the `size` argument of `fill_random` to `size_t`, add `res != -1` to this `if` condition and, for WIN32, check whether size exceeds `unsigned long`?

Reply:
Nice catch. I hate C integers.

Reply:
Argh,how do I check if `size` exceeds `ULONG_MAX` if the standard doesn't even tell me what's bigger `size_t` or `unsigned long`?.

What's the "right" way to do this here? :(

Reply:
What's wrong with `if (size > ULONG_MAX) return 0;`? Also, instead of checking `res != 1`, I think it would be better to check that `size <= SSIZE_MAX`.

Reply:
I was worried about what if `size_t` is smaller than `unsigned long`, the constant will overflow and the condition won't be the one it seems to be.
But I re-read the integer promotion rules and I think it checks out, if I don't explicitly convert them to a specific type then they'll be converted to the integer with the higher rank, preserving the right values.

Reply:
Yep, sounds correct what you say about the promotion rules.